planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,195 @@ import (
+// For SQL like `select max(id), min(id) from t;`, we could optimize it to the cartesianJoin result of the two queries above if `id` has an index.
 type maxMinEliminator struct {
 }
 
 func (a *maxMinEliminator) optimize(ctx context.Context, p LogicalPlan) (LogicalPlan, error) {
-	a.eliminateMaxMin(p)
-	return p, nil
+	return a.eliminateMaxMin(p), nil
 }
 
-// Try to convert max/min to Limit+Sort operators.
-func (a *maxMinEliminator) eliminateMaxMin(p LogicalPlan) {
-	// We don't need to guarantee that the child of it is a data source. This transformation won't be worse than previous.
-	if agg, ok := p.(*LogicalAggregation); ok {
-		// We only consider case with single max/min function.
-		if len(agg.AggFuncs) != 1 || len(agg.GroupByItems) != 0 {
-			return
+// composeAggsByInnerJoin composes the scalar aggregations by cartesianJoin.
+func (a *maxMinEliminator) composeAggsByInnerJoin(aggs []*LogicalAggregation) (plan LogicalPlan) {
+	plan = aggs[0]
+	sctx := plan.SCtx()
+	for i := 1; i < len(aggs); i++ {
+		join := LogicalJoin{JoinType: InnerJoin}.Init(sctx)
+		join.SetChildren(plan, aggs[i])
+		join.schema = buildLogicalJoinSchema(InnerJoin, join)
+		join.cartesianJoin = true
+		plan = join
+	}
+	return
+}
+
+// checkColCanUseIndex checks whether there is an accessPath satisfy the conditions:
+// 1. all of the selection's condition can be pushed down as accessConds of the path.
+// 2. the path can keep order for `col` after pushing down the conditions.
+func (a *maxMinEliminator) checkColCanUseIndex(plan LogicalPlan, col *expression.Column, conditions []expression.Expression) bool {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		conditions = append(conditions, p.Conditions...)
+		return a.checkColCanUseIndex(p.children[0], col, conditions)
+	case *DataSource:
+		// Check whether there is an accessPath can use index for col.
+		for _, path := range p.possibleAccessPaths {
+			if path.isTablePath {
+				// Since table path can contain accessConds of at most one column,
+				// we only need to check if all of the conditions can be pushed down as accessConds
+				// and `col` is the handle column.
+				if p.handleCol != nil && col.Equal(nil, p.handleCol) {
+					if _, filterConds := ranger.DetachCondsForColumn(p.ctx, conditions, col); len(filterConds) != 0 {
+						return false
+					}
+					return true
+				}
+			} else {
+				// For index paths, we have to check:
+				// 1. whether all of the conditions can be pushed down as accessConds.
+				// 2. whether the accessPath can satisfy the order property of `col` with these accessConds.
+				result, err := ranger.DetachCondAndBuildRangeForIndex(p.ctx, conditions, path.fullIdxCols, path.fullIdxColLens)
+				if err != nil || len(result.RemainedConds) != 0 {
+					continue
+				}
+				for i := 0; i <= result.EqCondCount; i++ {
+					if i < len(path.fullIdxCols) && col.Equal(nil, path.fullIdxCols[i]) {
+						return true
```

Comment:
Shall we prune other access paths here?

Reply:
We can do so, but I'm afraid it may increase the coupling degree which makes it hard to maintain.

Reply:
Why would it increase coupling? Oh, there might exist multiple indexes which satisfy the requirement, we'd better keep these access paths to be decided by physical optimization.